### PR TITLE
Fixed Maven/Gradle code analysis uploading empty artifacts.

### PR DESCRIPTION
**Pull request by HSAR**

### CODE REVIEW

File: Tasks/Gradle/CodeAnalysis/Common/CodeAnalysisOrchestrator.ts [hunk 1]
```diff
@@ -52,8 +52,8 @@ export class CodeAnalysisOrchestrator {
 
             let resultPublisher = new CodeAnalysisResultPublisher(analysisResults, stagingDir);
 
-            resultPublisher.uploadBuildSummary();
-            resultPublisher.uploadArtifacts(buildNumber);
+            var uploadedArtifacts:number = resultPublisher.uploadArtifacts(buildNumber);
+            resultPublisher.uploadBuildSummary(uploadedArtifacts);
```
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/Common/CodeAnalysisOrchestrator.ts]: uploadBuildSummary  needs to know whether or not we actually uploaded any artifacts.

File: Tasks/Gradle/CodeAnalysis/Common/CodeAnalysisResultPublisher.ts [hunk 2]
```diff
@@ -24,8 +24,20 @@ export class CodeAnalysisResultPublisher {
...
+            tl.debug("Skipping artifact upload: No analysis results");
+            return;
+        }
+
+        // If there are no files to upload, return
+        var analysisResultsWithFiles:AnalysisResult[] = this.analysisResults.filter(
+            (analysisResult:AnalysisResult) => {
+                return analysisResult.resultFiles != undefined && analysisResult.resultFiles != null && analysisResult.resultFiles.length > 0;
+        });
+        if (analysisResultsWithFiles.length === 0) {
```
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/Common/CodeAnalysisResultPublisher.ts]: This is the core of the fix - if no analysisResult objects had files to upload, then return early.

File: Tasks/Gradle/CodeAnalysis/Common/CodeAnalysisResultPublisher.ts [hunk 3]
```diff
@@ -24,8 +24,20 @@ export class CodeAnalysisResultPublisher {
...
-    public uploadArtifacts(prefix: string):void {
+    public uploadArtifacts(prefix: string):number {
+        // If there are no results to upload, return
         if (this.analysisResults.length === 0) {
+            tl.debug("Skipping artifact upload: No analysis results");
+            return;
+        }
+
+        // If there are no files to upload, return
+        var analysisResultsWithFiles:AnalysisResult[] = this.analysisResults.filter(
```
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/Common/CodeAnalysisResultPublisher.ts]: Using Javascript feature here - _filter()_ creates a new array with only the elements for which the callback returns true. The callback passed below does a null/undefined check for safety, then returns true if there were files to upload.

File: Tasks/Gradle/CodeAnalysis/Common/CodeAnalysisResultPublisher.ts [hunk 4]
```diff
@@ -111,7 +124,7 @@ export class CodeAnalysisResultPublisher {
             }
         }
 
-        if (buildSummaryLines.length > 0) {
+        if (buildSummaryLines.length > 0 && uploadedArtifacts > 0) { // Do not print this last line if there were no results uploaded
             buildSummaryLines.push('');
             buildSummaryLines.push('Code analysis results can be found in the \'Artifacts\' tab.');
```
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/Common/CodeAnalysisResultPublisher.ts]: Fit and finish: there's no point telling the user to find code analysis results in the artifacts tab if, in fact, no artifacts were uploaded.

File: Tasks/Gradle/CodeAnalysis/SonarQube/vsts-server-utils.ts [hunk 5]
```diff
@@ -60,14 +60,14 @@ export class VstsServerUtils {
      * @returns {Promise<void>} Promise resolved when action completes.
      */
     public static processSonarQubeBuildSummary(sqRunSettings:SonarQubeRunSettings, sqMetrics:SonarQubeMetrics):Q.Promise<void> {
-    // During a pull request build, data necessary to create SQRunSettings is not available
-    if (sqRunSettings == null || VstsServerUtils.isPrBuild()) {
-        console.log(tl.loc('sqAnalysis_IsPullRequest_SkippingBuildSummary'));
-        return Q.when<void>(null);
-    }
+        // During a pull request build, data necessary to create SQRunSettings is not available
```
Comment (HSAR) [inline Tasks/Gradle/CodeAnalysis/SonarQube/vsts-server-utils.ts]: Unrelated indentation fix.

**Comment by HSAR** [comment]
This PR is missing test work - add before merging.

**Comment by bgavrilMS** [comment]
LGTM

File: Tests/L0/Gradle/_suite.ts [hunk 6]
```diff
@@ -1265,17 +1250,74 @@ describe('gradle Suite', function () {
...
+                assert(tr.stdout.indexOf('##vso[artifact.upload artifactname=Code Analysis Results;]') < 0,
+                    'should not have uploaded a code analysis build artifact');
+
+                assertCodeAnalysisBuildSummaryContains(testStgDir, 'Checkstyle found no violations.');
+                assertCodeAnalysisBuildSummaryContains(testStgDir, 'PMD found no violations.');
+                assertCodeAnalysisBuildSummaryContains(testStgDir, 'FindBugs found no violations.');
+
+                // The .codeAnalysis dir should have been created to store the build summary, but not the report dirs
+                var codeAnalysisStgDir: string = path.join(testStgDir, '.codeAnalysis');
+                assertFileDoesNotExistInDir(codeAnalysisStgDir, 'CA');
```
Comment (duncanpMS) [inline Tests/L0/Gradle/_suite.ts]: Does this method assert if the directory does not exist?

File: Tests/L0/Maven/_suite.ts [hunk 7]
```diff
@@ -1688,13 +1737,12 @@ describe('Maven Suite', function () {
         cleanTempDirsForCodeAnalysisTests();
     });
 
-    it('Maven with code analysis - Executes and uploads results for all enabled tools', function (done) {
+    it('Maven with code analysis - Executes and uploads discovered results even if the tools were not enabled', function (done) {
```
Comment (duncanpMS) [inline Tests/L0/Maven/_suite.ts]: Is this a change of behaviour, or just fixing an inaccurate comment?

File: Tests/L0/Gradle/_suite.ts [hunk 8]
```diff
@@ -1265,17 +1250,74 @@ describe('gradle Suite', function () {
...
+                console.log(tr.stderr);
+                console.log(err);
+                done(err);
+            });
+
+        // Clean up
+        cleanTempDirsForCodeAnalysisTests();
+    });
+
+    it('Gradle with code analysis - Does not upload artifacts if code analysis reports were empty', function (done) {
```
Comment (duncanpMS) [inline Tests/L0/Gradle/_suite.ts]: Is there a test for the case where tool X produces analysis reports but tool Y doesn't?

File: Tests/L0/Maven/_suite.ts (hunk 7 above)
Comment (HSAR) [inline Tests/L0/Maven/_suite.ts]: Fixing an inaccurate comment.

File: Tests/L0/Gradle/_suite.ts (hunk 8 above)
Comment (HSAR) [inline Tests/L0/Gradle/_suite.ts]: https://github.com/Microsoft/vsts-tasks/pull/2746/files/6b6f629a680cfd457431d0eece5b29a6e326e68f#diff-260cb51b9bd3175ad8b53a45ae5c59caR1222 

The above test asserts that when only one tool is enabled and three empty reports are found, that an artifact is uploaded and the build summary contains only the line for the tool which is enabled.

File: Tests/L0/Gradle/_suite.ts (hunk 6 above)
Comment (HSAR) [inline Tests/L0/Gradle/_suite.ts]: Indeed, "CA" in this line is a folder.

Comment (duncanpMS) [inline Tests/L0/Gradle/_suite.ts]: I meant the parent directory i.e. _codeAnalysisStgDir_. The comment in line 1318 suggests that the parent directory should exist - is its existence being checked?

Comment (HSAR) [inline Tests/L0/Gradle/_suite.ts]: I see. Yes, the test method asserts both that the base directory exists, and that it contains the specific file or folder.

**Comment by HSAR** [comment]
I'll be conducting a small edit to the commits to seperate out some of the changes, in case we need to cherry-pick just the essential comjmit later.
